Modify and/or eliminate unused MPI bookkeepings that cause race conditions.
--- src/sys/objects/inherit.c
+++ src/sys/objects/inherit.c
@@ -3,6 +3,7 @@
 */
 #include <petsc/private/petscimpl.h> /*I   "petscsys.h"    I*/
 #include <petscviewer.h>
+#include <stdatomic.h>

 #if defined(PETSC_USE_LOG)
 PETSC_INTERN PetscObject *PetscObjects;
@@ -24,8 +25,8 @@ PETSC_EXTERN PetscErrorCode PetscObjectQueryFunction_Petsc(PetscObject, const ch

 PetscObjectId PetscObjectNewId_Internal(void)
 {
-  static PetscObjectId idcnt = 1;
-  return idcnt++;
+  static atomic_int atomic_idcnt = 1;
+  return atomic_fetch_add(&atomic_idcnt, 1);
 }

 /*
@@ -55,14 +56,6 @@ PetscErrorCode PetscHeaderCreate_Private(PetscObject h, PetscClassId classid, co

   PetscCall(PetscCommDuplicate(comm, &h->comm, &h->tag));

-  /* Increment and store current object creation index */
-  PetscCallMPI(MPI_Comm_get_attr(h->comm, Petsc_CreationIdx_keyval, &get_tmp, &flg));
-  if (flg) {
-    cidx    = (PetscInt64 *)get_tmp;
-    h->cidx = (*cidx)++;
-    PetscCallMPI(MPI_Comm_set_attr(h->comm, Petsc_CreationIdx_keyval, cidx));
-  } else SETERRQ(h->comm, PETSC_ERR_ARG_CORRUPT, "MPI_Comm does not have an object creation index");
-
 #if defined(PETSC_USE_LOG)
   /* Keep a record of object created */
   if (PetscObjectsLog) {
